library/alloc/src/vec/mod.rs

```diff
@@ -2274,6 +2274,23 @@ impl<T: Clone, A: Allocator> Vec<T, A> {
     }
 }
 
+impl<T, A: Allocator, const N: usize> Vec<[T; N], A> {
+    #[unstable(feature = "slice_flatten", issue = "none")]
+    pub fn into_flattened(self) -> Vec<T, A> {
+        let (ptr, len, cap, alloc) = self.into_raw_parts_with_alloc();
+        let new_len = if crate::mem::size_of::<T>() == 0 {
+            self.len().checked_mul(N).expect("vec len overflow")
+        } else {
+            // SAFETY: `self.len() * N` cannot overflow because the allocation
+            // is already in the address space.
+            unsafe { self.len().unchecked_mul(N) }
+        };
+        // SAFETY: Arrays are contiguous, so the layout of the allocation has
+        // `len * N` elements,
+        unsafe { Vec::<T, A>::from_raw_parts_in(ptr.cast(), new_len, cap, alloc) }
```

Comment:
This is very much incorrect right now, as the capacity is also in units of items, not bytes, so also needs to be multiplied.

I'll let you think up the exact way to handle capacities in the ZST and non-ZST cases; let me know if you need help though.

Reply:
Oops, completely forgot about capacity. Pushing fix now